File: backend/app/models/incident.py
from datetime import datetime, timezone
import json
import enum
from typing import Optional, List, Dict, Any
from sqlalchemy import String, Text, Integer, DateTime, ForeignKey, Enum as SQLEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from backend.app.core.database import Base

class IncidentStatus(str, enum.Enum):
    NEW = "NEW"
    INVESTIGATING = "INVESTIGATING"
    DIAGNOSED = "DIAGNOSED"
    ACTION_REQUIRED = "ACTION_REQUIRED"
    REMEDIATING = "REMEDIATING"
    RESOLVED = "RESOLVED"
    ESCALATED = "ESCALATED"

class IncidentSeverity(str, enum.Enum):
    CRITICAL = "critical"
    WARNING = "warning"
    INFO = "info"

def utc_now() -> datetime:
    return datetime.now(timezone.utc)

class Incident(Base):
    __tablename__ = "incidents"

    id: Mapped[str] = mapped_column(String(64), primary_key=True, index=True)
    fingerprint: Mapped[str] = mapped_column(String(128), index=True)
    title: Mapped[str] = mapped_column(String(256))
    service: Mapped[str] = mapped_column(String(128), index=True)
    namespace: Mapped[str] = mapped_column(String(128), default="default", index=True)
    alert_name: Mapped[str] = mapped_column(String(128), index=True)
    severity: Mapped[str] = mapped_column(String(32), default=IncidentSeverity.WARNING.value, index=True)
    status: Mapped[str] = mapped_column(String(32), default=IncidentStatus.NEW.value, index=True)
    
    summary: Mapped[str] = mapped_column(Text, default="")
    description: Mapped[str] = mapped_column(Text, default="")
    firing_count: Mapped[int] = mapped_column(Integer, default=1)
    
    first_seen_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now)
    last_seen_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now)
    resolved_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)

    labels_json: Mapped[str] = mapped_column(Text, default="{}")
    annotations_json: Mapped[str] = mapped_column(Text, default="{}")
    evidence_json: Mapped[str] = mapped_column(Text, default="{}")
    diagnosis_json: Mapped[str] = mapped_column(Text, default="{}")
    remediation_json: Mapped[str] = mapped_column(Text, default="{}")

    events: Mapped[List["IncidentEvent"]] = relationship(
        "IncidentEvent", back_populates="incident", cascade="all, delete-orphan", order_by="IncidentEvent.created_at"
    )

    @property
    def labels(self) -> Dict[str, Any]:
        return json.loads(self.labels_json) if self.labels_json else {}

    @labels.setter
    def labels(self, val: Dict[str, Any]):
        self.labels_json = json.dumps(val)

    @property
    def annotations(self) -> Dict[str, Any]:
        return json.loads(self.annotations_json) if self.annotations_json else {}

    @annotations.setter
    def annotations(self, val: Dict[str, Any]):
        self.annotations_json = json.dumps(val)

    @property
    def evidence(self) -> Dict[str, Any]:
        return json.loads(self.evidence_json) if self.evidence_json else {}

    @evidence.setter
    def evidence(self, val: Dict[str, Any]):
        self.evidence_json = json.dumps(val)

    @property
    def diagnosis(self) -> Dict[str, Any]:
        return json.loads(self.diagnosis_json) if self.diagnosis_json else {}

    @diagnosis.setter
    def diagnosis(self, val: Dict[str, Any]):
        self.diagnosis_json = json.dumps(val)

    @property
    def remediation(self) -> Dict[str, Any]:
        return json.loads(self.remediation_json) if self.remediation_json else {}

    @remediation.setter
    def remediation(self, val: Dict[str, Any]):
        self.remediation_json = json.dumps(val)

class IncidentEvent(Base):
    __tablename__ = "incident_events"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    incident_id: Mapped[str] = mapped_column(String(64), ForeignKey("incidents.id"), index=True)
    event_type: Mapped[str] = mapped_column(String(64), index=True)
    message: Mapped[str] = mapped_column(Text)
    payload_json: Mapped[str] = mapped_column(Text, default="{}")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now)

    incident: Mapped["Incident"] = relationship("Incident", back_populates="events")

    @property
    def payload(self) -> Dict[str, Any]:
        return json.loads(self.payload_json) if self.payload_json else {}

    @payload.setter
    def payload(self, val: Dict[str, Any]):
        self.payload_json = json.dumps(val)

File: frontend/src/types/incident.ts
// Core incident and diagnostic types shared across all components

export type IncidentStatus =
  | "NEW"
  | "INVESTIGATING"
  | "DIAGNOSED"
  | "ACTION_REQUIRED"
  | "REMEDIATING"
  | "RESOLVED"
  | "ESCALATED";

export type IncidentSeverity = "critical" | "warning" | "info";
export type RiskTier = "TIER_1_LOW" | "TIER_2_MEDIUM" | "TIER_3_HIGH";

export interface IncidentEvent {
  id: number;
  incident_id: string;
  event_type: string;
  message: string;
  payload: Record<string, unknown>;
  created_at: string;
}

export interface GoldenSignals {
  request_rate_rps: number;
  error_rate_pct: number;
  latency_p50_ms: number;
  latency_p95_ms: number;
  latency_p99_ms: number;
  cpu_saturation_pct: number;
  memory_saturation_pct: number;
}

export interface SignalTimeline {
  timestamps: string[];
  error_rate: number[];
  latency_p99: number[];
  memory_saturation: number[];
}

export interface DeploymentMarker {
  timestamp: string;
  revision: string;
  message: string;
}

export interface GoldenSignalsData {
  service: string;
  namespace: string;
  timestamp: string;
  signals: GoldenSignals;
  timeline: SignalTimeline;
  deployment_marker?: DeploymentMarker;
}

export interface PodInfo {
  name: string;
  service: string;
  namespace: string;
  status: string;
  ready: boolean;
  restart_count: number;
  exit_code: number;
  reason: string | null;
  node: string;
  started_at: string;
}

export interface ClusterEvent {
  type: string;
  reason: string;
  object: string;
  message: string;
  timestamp: string;
}

export interface LogEntry {
  pod: string;
  message: string;
}

export interface LogData {
  service: string;
  namespace: string;
  total_lines_scanned: number;
  error_count: number;
  errors: LogEntry[];
  raw_logs: LogEntry[];
}

export interface DeploymentInfo {
  name: string;
  namespace: string;
  replicas: number;
  ready_replicas: number;
  image: string;
  revision: number;
  history: Array<{ revision: number; image: string; updated_at: string }>;
}

export interface InvestigationEvidence {
  service: string;
  namespace: string;
  collected_at: string;
  golden_signals: GoldenSignalsData;
  pods: PodInfo[];
  events: ClusterEvent[];
  deployment: DeploymentInfo;
  logs: LogData;
  anomalies: string[];
  health_score: number;
}

export interface Diagnosis {
  root_cause: string;
  confidence: number;
  supporting_evidence: string[];
  recommended_action: string;
  action_type: string;
  risk_tier: RiskTier;
  requires_escalation: boolean;
  diagnosed_at: string;
}

export interface RemediationAction {
  action_type: string;
  name: string;
  description: string;
  risk_tier: RiskTier;
  target_resource: string;
  parameters: Record<string, unknown>;
  requires_human_approval: boolean;
  is_auto_executable: boolean;
  diff_preview: string;
}

export interface Incident {
  id: string;
  fingerprint: string;
  title: string;
  service: string;
  namespace: string;
  alert_name: string;
  severity: IncidentSeverity;
  status: IncidentStatus;
  summary: string;
  description: string;
  firing_count: number;
  first_seen_at: string;
  last_seen_at: string;
  resolved_at: string | null;
  labels: Record<string, string>;
  annotations: Record<string, string>;
  evidence: InvestigationEvidence | Record<string, never>;
  diagnosis: Diagnosis | Record<string, never>;
  remediation: RemediationAction | Record<string, never>;
  events: IncidentEvent[];
}

export interface SSEMessage {
  event: string;
  data: Record<string, unknown>;
  timestamp: string;
}
